### ISSUE TITLE
l4d2_nobhaps plugin flaws

### ISSUE THREAD

**Issue by Derpduck** [SirPlease/L4D2-Competitive-Rework#290]
This plugin has a number of flaws that affects all players' movement even in situations where it isn't supposed to (i.e. they aren't trying to bhop/gain speed).

1. Speed cap is applied when landing on the ground and not when starting a jump, which causes it to cap speeds even when players aren't attempting to bhop or chain jumps at all, and overrides the game's natural ground friction.
2. Maximum velocity is stored from the previous time the player was on the ground, any case where their maximum velocity changes while mid-air the bhop check will be based on the old max speed.
    * This means that every rocket spawn is able to get one uncapped bhop, rather than being capped to their spawned move speed.
3. Landing a jump on objects at a certain height will briefly force players to crouch, which can cause the plugin to slow the player's speed to 75 on the jump after.
4. Landing from a high fall as an infected forces you to crouch, resulting in a speed cap of 75 from the next jump.
    * This can be seen in combination with no. 2 where players can get a full speed hop but then will always get capped to 75 speed on the next one, even though there was a full jump between the crouch.
5. Being crouched and attempting to hop will instantly set speed to 75, even if the player uncrouches mid-air, even though with the default physics you should be able to preserve the move speed you had when you first jumped. Speed should be capped to your maximum possible walking speed at the time to prevent this.

**Comment by A1mDev** [comment]
Yes, indeed he constantly limits the speed, even if it is higher due to the fact that the player is still in the jump. You can verify this by including messages in the chat if the plugin fix works. And perhaps the problem indicated here is related to this plugin - #235 (difficult to check). However, actually changing the speed with help TeleportEntity() may be bad code for game movement. Another option is to block the jump at the right moment, but from the client's point of view it looks bad, apparently then the client code is too different, but you can try. And another option to improve the accuracy of this plugin (limit speed if the player is doing a clean bhop and not just jumping).

**Comment by A1mDev** [comment]
Although you may need to just decrease the z-axis speed, then this will work fine =)

**Comment by robex** [comment]
You could check for a perfect bhop the same way SMAC does it (TLDR; check if you jumped immediately at the tick after you touched the ground):
https://github.com/Silenci0/SMAC/blob/master/addons/sourcemod/scripting/smac_autotrigger.sp#L95

And if this happens then limit speed by maybe, instead of teleporting, setting the velocity vector manually but capping the speed to max walking speed for the player's HP. I'm not entirely sure if this would work though.

This would break the 1-hop rule for rocket spawns which I think is a nice mechanic to have, but you could keep it by allowing 1 single hop for SI.

**Comment by robex** [comment]
addressed in PR #304

**Comment by robex** [comment]
So, it seems like I messed something up with #304. Perhaps its OnPlayerRunCmd not firing accurately, but it seems like on SirPlease servers you can now gain speed while jumping somewhat unrestrictedly (https://www.twitch.tv/videos/1067307507?t=0h48m55s). I have been able to reproduce it on my local server only once, so for now I suggest reverting back to the old plugin but with the crouching speed issue fixed until further investigation, I've made a new PR for it (#306).

**Comment by SirPlease** [comment]
I'll leave this open for now until there's confirmation we've got all of these covered.
